free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py

```diff
@@ -0,0 +1,390 @@
+
+        self.tf_handler = Nav2TfHandler(
+            self.name,
+            self.zenoh_session,
+            self.tf_buffer,
+            self.node
+        )
+
+        def _battery_state_callback(sample: zenoh.Sample):
+            battery_state = SensorMsgs_BatteryState.deserialize(
+                sample.payload.to_bytes()
+            )
+            self.battery_soc = battery_state.percentage
+
+        self.battery_state_sub = self.zenoh_session.declare_subscriber(
+            namespacify('battery_state', name),
+            _battery_state_callback
+        )
+
+    def get_battery_soc(self) -> float:
+        return self.battery_soc
+
+    def get_pose(self) -> Annotated[list[float], 3] | None:
+        transform = self.tf_handler.get_transform()
+        if transform is None:
+            error_message = \
+                f'Failed to update robot [{self.name}]: Unable to get ' \
+                f'transform between base_footprint and map'
+            self.node.get_logger().info(error_message)
+            return None
+
+        orientation = euler_from_quaternion([
+            transform.transform.rotation.x,
+            transform.transform.rotation.y,
+            transform.transform.rotation.z,
+            transform.transform.rotation.w
+        ])
+        robot_pose = [
+            transform.transform.translation.x,
+            transform.transform.translation.y,
+            orientation[2]
+        ]
+        return robot_pose
+
+    def _is_navigation_done(self) -> bool:
+        if self.nav_goal_id is None:
+            return True
+
+        req = NavigateToPose_GetResult_Request(goal_id=self.nav_goal_id)
+        # TODO(ac): parameterize the service call timeout
+        replies = self.zenoh_session.get(
+            namespacify('navigate_to_pose/_action/get_result', self.name),
+            payload=req.serialize(),
+            # timeout=0.5
+        )
+        for reply in replies:
+            try:
+                # Deserialize the response
+                rep = NavigateToPose_GetResult_Response.deserialize(
+                    reply.ok.payload.to_bytes()
+                )
+                self.node.get_logger().debug(f'Result: {rep.status}')
+                match rep.status:
+                    case GoalStatus.STATUS_EXECUTING.value | \
+                         GoalStatus.STATUS_ACCEPTED.value | \
+                         GoalStatus.STATUS_CANCELING.value:
+                        return False
+                    case GoalStatus.STATUS_SUCCEEDED.value:
+                        self.node.get_logger().info(
+                            f'Navigation goal {self.nav_goal_id} reached'
+                        )
+                        return True
+                    case GoalStatus.STATUS_CANCELED.value:
+                        self.node.get_logger().info(
+                            f'Navigation goal {self.nav_goal_id} was cancelled'
+                        )
+                        return True
```

Comment:
I believe this case is handled under `stop()`, so we can probably leave `CANCELED` scenarios out. When RMF requests for a robot to stop, leading to all goal requests being canceled, `self.execution` and `self.nav_goal_id` are already set to `None` inside the `stop()` and `_handle_stop_navigation()` callbacks respectively. On the other hand, `is_navigation_done()` is only ever called when both `self.execution` and `self.nav_goal_id` are not `None`, so this case would be redundant here.

Suggesting to leave it out because it'd be good to have a single source of truth/logic for goal canceled cases, especially when we are returning `True` and telling RMF that we have successfully completed a navigation task. Feel free to let me know if I'm misunderstanding the usage.

Reply:
So far the case doesn't do anything other than tell the robot adapter that navigation is done, there was a race condition I encountered while testing this behavior, that was "fixed" with https://github.com/open-rmf/free_fleet/pull/145/commits/9cf376a3fb9064ea43b95135c43443236903fb06#diff-6f7a8ca807eab341270b8ed8d8073dcea66db64ff693dd774d681225f2f526ab

Currently this code block doesn't do anything other than produce a log, and report that navigation is done. The alternative would be more hairy, as it will start going into the logic for replanning.